pkg/vsphere/disk/disk_manager.go

```diff
@@ -275,24 +281,38 @@ func (m *Manager) Detach(op trace.Operation, d *VirtualDisk) error {
 		return errors.Trace(err)
 	}
 
-	spec := types.VirtualMachineConfigSpec{}
-
-	disk, err := findDisk(op, m.vm, d.DatastoreURI)
+	disk, err := findDiskByFilename(op, m.vm, d.DatastoreURI)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
+	if err = m.detach(op, disk); err != nil {
+		op.Errorf("detach for %s failed with %s", d.DevicePath, errors.ErrorStack(err))
+		return errors.Trace(err)
+	}
+
+	select {
```

Comment:
Blocking forever isn't acceptable. I'd rather code defensively in that instance. To be fair, there should never be an instance where the channel can't be drained because we have a lock around the push.